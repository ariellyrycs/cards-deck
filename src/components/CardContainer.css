
.card-container {
    width: 55rem;
    height: 35rem;
    display: flex;
    justify-content: space-between;
}

.card-container > * {
    width: 25rem;
    border: 1px solid blue;
    border-radius: 10px;
}

.unrevealed, .revealed {
    position: relative;
    perspective: 100px;
    perspective-origin: 105% 125%;
    overflow: hidden;
}

.card {
    display: block;
    width: 20rem;
    height: 28rem;
    position: absolute;
    left: 50%;
    top: 50%;
    background: #fff;
    background-color: transparent;
    cursor: pointer;
    opacity: 1;
    transition: all 0.5s;
    perspective: 1000px;
}

.flip-card-inner {
    width: 100%;
    height: 100%;
    position: relative;
    text-align: center;
    transition: transform 0.6s;
    transform-style: preserve-3d;
    box-shadow: 0 4px 8px 0 rgba(0,0,0,0.2);
}

.card.flipped .flip-card-inner {
    transform: rotateY(180deg);
}

.flip-card-front, .flip-card-back {
    position: absolute;
    width: 100%;
    height: 100%;
    -webkit-backface-visibility: hidden;
    backface-visibility: hidden;
}

.flip-card-front {
    color: black;
    background-image: url('../back-card.jpg');
    background-position: center;
}

.flip-card-back {
    background-color: white;
    color: black;
    transform: rotateY(180deg);
    display: flex;
    justify-content: center;
    align-items: center;
    font-size: 13rem;
}
