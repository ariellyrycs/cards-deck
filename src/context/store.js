
import React, { useReducer, createContext, useContext } from 'react';
const Context = createContext();



const useAppContext = () => useContext(Context);

let randomizeIndexes = arr => {
    var len = arr.length;
    for(let i = 0; i < len; i += 1) {
        let o = Math.floor(Math.random() * len);
        [arr[i], arr[o]] = [arr[o], arr[i]];
    }
    return arr;
};


const initialState = {
    cards: {
        considered: randomizeIndexes([{
            id: 0,
            value: 1,
            revealed: false
        }, {
            id: 1,
            value: 2,
            revealed: false
        }, {
            id: 2,
            value: 3,
            revealed: false
        }, {
            id: 3,
            value: 4,
            revealed: false
        }, {
            id: 4,
            value: 5,
            revealed: false
        }, {
            id: 5,
            value: 6,
            revealed: false
        }, {
            id: 6,
            value: 7,
            revealed: false
        }, {
            id: 7,
            value: 8,
            revealed: false
        }, {
            id: 8,
            value: 9,
            revealed: false
        }]),
        unconsidered: []
    } 
};

const mainReducer = (state, action) => {
    const cards = state.cards;
    let considered;
    
    switch(action.type) {
        case 'REVEAL_TOP':
            considered = cards.considered.slice(0);
            considered[considered.length - 1].revealed = true;
            
            return {
                ...state,
                cards: {
                    ...cards,
                    considered
                }
            };
        case 'POP_CARD':
            considered = cards.considered.slice(0);
            return {
                ...state,
                cards: {
                    ...cards,
                    considered,
                    unconsidered: [...cards.unconsidered, considered.pop()]
                }
            };
        default:
            return state;
    }
};

function ContextProvider(props) {
    const [state, dispatch] = useReducer(mainReducer, initialState);

    return (
        <Context.Provider
            value={{state, dispatch}}>
            {props.children}
        </Context.Provider>
    )
}

export {ContextProvider, Context, useAppContext};