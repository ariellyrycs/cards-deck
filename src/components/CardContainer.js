
import React from 'react';
import './CardContainer.css';
import { useAppContext } from '../context/store';

export default function CardContainer() {
    const { state, dispatch } = useAppContext();
    return (
        <div className="card-container">
            <div className="unrevealed">
                {state.cards.considered.map((card, i) => (
                    <div
                        className={`card flip-card ${card.revealed ? 'flipped': ''}`}
                        key={i}
                        style={{
                            transform: `translate3d(-50%, -50%, ${1 * i}px)`,
                            zIndex: `-(arr.length - i - 1)`
                        }}
                        onClick={() => dispatch({type: card.revealed ? 'POP_CARD': 'REVEAL_TOP'})}
                        >
                        <div className="flip-card-inner">
                            <div className="flip-card-front"></div>
                            <div className="flip-card-back">
                                {card.value}
                            </div>
                        </div>
                    </div>
                ))}
            </div>
            <div className="revealed">
                {state.cards.unconsidered.map((card, i) => (
                    <div
                        className="card flip-card flipped"
                        key={i}
                        style={{
                            transform: `translate3d(-50%, -50%, ${1 * i}px)`,
                            zIndex: `-(arr.length - i - 1)`
                        }}
                        >
                        <div className="flip-card-inner">
                            <div className="flip-card-front"></div>
                            <div className="flip-card-back">
                                {card.value}
                            </div>
                        </div>
                    </div>
                ))}
            </div>
        </div>
    );
};