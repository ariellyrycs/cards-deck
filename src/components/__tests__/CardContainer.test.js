import CardContainer from '../CardContainer';
import { shallow, configure} from 'enzyme';
import toJson from 'enzyme-to-json';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';

import * as Store from '../../context/store.js';


configure({ adapter: new Adapter() });

describe('<CardContainer />', () => {
    test('it should mock context', () => {
        const contextValues = {
            state: {
                cards: {
                    considered: [{
                        id: 0,
                        value: 1,
                        revealed: false
                    }, {
                        id: 1,
                        value: 2,
                        revealed: true
                    }],
                    unconsidered: []
                }
            }
        };

        jest
            .spyOn(Store, 'useAppContext')
            .mockImplementation(() => contextValues);
            
        const wrapper = shallow(<CardContainer />);
        expect(toJson(wrapper)).toMatchSnapshot();
    });
});