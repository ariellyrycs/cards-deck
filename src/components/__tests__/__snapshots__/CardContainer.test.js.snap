// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CardContainer /> it should mock context 1`] = `
<div
  className="card-container"
>
  <div
    className="unrevealed"
  >
    <div
      className="card flip-card "
      key="0"
      onClick={[Function]}
      style={
        Object {
          "transform": "translate3d(-50%, -50%, 0px)",
          "zIndex": "-(arr.length - i - 1)",
        }
      }
    >
      <div
        className="flip-card-inner"
      >
        <div
          className="flip-card-front"
        />
        <div
          className="flip-card-back"
        >
          1
        </div>
      </div>
    </div>
    <div
      className="card flip-card flipped"
      key="1"
      onClick={[Function]}
      style={
        Object {
          "transform": "translate3d(-50%, -50%, 1px)",
          "zIndex": "-(arr.length - i - 1)",
        }
      }
    >
      <div
        className="flip-card-inner"
      >
        <div
          className="flip-card-front"
        />
        <div
          className="flip-card-back"
        >
          2
        </div>
      </div>
    </div>
  </div>
  <div
    className="revealed"
  />
</div>
`;
